packages/flutter/lib/src/gestures/pointer_data_resampler.dart

```diff
@@ -0,0 +1,199 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+// @dart = 2.8
+
+import 'dart:ui' as ui;
+
+/// Class for pointer data resampling.
+class PointerDataResampler {
+  // Data enqueued for processing.
+  final _queuedData = <ui.PointerData>[];
+
+  /// Pointer state required for resampling.
+  ui.PointerData _last;
+  ui.PointerData _next;
+  var _position = ui.Offset(0.0, 0.0);
+  var _isTracked = false;
+  var _isDown = false;
+  var _pointerIdentifier = 0;
+
+  ui.PointerData _clone(
+      ui.PointerData data,
+      ui.PointerChange change,
+      ui.Offset position,
+      ui.Offset delta,
+      int pointerIdentifier,
+      Duration timeStamp) {
+    return ui.PointerData(
+        buttons: data.buttons,
+        device: data.device,
+        timeStamp: timeStamp,
+        change: change,
+        kind: data.kind,
+        physicalDeltaX: delta.dx,
+        physicalDeltaY: delta.dy,
+        physicalX: position.dx,
+        physicalY: position.dy,
+        pointerIdentifier: pointerIdentifier,
+        synthesized: data.synthesized);
+  }
+
+  ui.Offset _positionAt(Duration sampleTime) {
+    // Use `next` position by default.
+    var x = _next.physicalX;
+    var y = _next.physicalY;
+
+    // Resample if `next` time stamp is past `sampleTime`.
+    if (_next.timeStamp > sampleTime && _next.timeStamp > _last.timeStamp) {
+      final interval =
+          (_next.timeStamp - _last.timeStamp).inMicroseconds.toDouble();
+      final scalar =
+          (sampleTime - _last.timeStamp).inMicroseconds.toDouble() / interval;
+      x = _last.physicalX + (_next.physicalX - _last.physicalX) * scalar;
+      y = _last.physicalY + (_next.physicalY - _last.physicalY) * scalar;
+    }
+
+    return ui.Offset(x, y);
+  }
+
+  void _processPointerData(Duration sampleTime) {
```

Comment:
We need to keep something like the previous code now that we're using a queue. I made it more efficient by stopping when the first data with time stamp greater than sample time is found. The amount of data processed here should always be low. The only time we might set last/next multiple times is when input frequency is higher than refresh rate and it would need to be extremely high for this processing to become a problem.

Reply:
Sounds good. Then maybe it suffices to add a comment like "Find _next to be the data with the smallest timeStamp that's no less than sampleTime, and the _last to be the previous event before _next." for easier understanding.